File: problems/problems_LCR_063/solution.go
package problemLCR_063

import (
	"encoding/json"
	"log"
	"strings"
)

type Trie struct {
	children [26]*Trie
	isEnd    bool
}

func Constructor() Trie {
	return Trie{}
}

func (t *Trie) Insert(word string) {
	node := t
	for _, ch := range word {
		ch -= 'a'
		if node.children[ch] == nil {
			node.children[ch] = &Trie{}
		}
		node = node.children[ch]
	}
	node.isEnd = true
}

func (t *Trie) FindPrefix(word string) string {
	node := t
	for i, ch := range word {
		ch -= 'a'
		if node.isEnd {
			return word[:i]
		}
		if node.children[ch] == nil {
			break
		}
		node = node.children[ch]
	}
	return word
}

func replaceWords(dictionary []string, sentence string) string {
	trie := Constructor()
	for _, word := range dictionary {
		trie.Insert(word)
	}
	words := strings.Split(sentence, " ")
	for i, word := range words {
		words[i] = trie.FindPrefix(word)
	}
	return strings.Join(words, " ")
}

func Solve(inputJsonValues string) interface{} {
	inputValues := strings.Split(inputJsonValues, "\n")
	var dictionary []string
	var sentence string

	if err := json.Unmarshal([]byte(inputValues[0]), &dictionary); err != nil {
		log.Fatal(err)
	}
	if err := json.Unmarshal([]byte(inputValues[1]), &sentence); err != nil {
		log.Fatal(err)
	}

	return replaceWords(dictionary, sentence)
}

File: problems/problems_LCR_063/Solution.java
package problems.problems_LCR_063;

import com.alibaba.fastjson.JSON;
import java.util.*;
import qubhjava.BaseSolution;


public class Solution extends BaseSolution {
    private TrieNode root;
    class TrieNode {
        TrieNode[] children;
        boolean isEnd;

        public TrieNode() {
            children = new TrieNode[26];
            isEnd = false;
        }
    }

    private void insert(String word) {
        TrieNode node = root;
        for (char c : word.toCharArray()) {
            int index = c - 'a';
            if (node.children[index] == null) {
                node.children[index] = new TrieNode();
            }
            node = node.children[index];
        }
        node.isEnd = true;
    }

    private String search(String word) {
        TrieNode node = root;
        StringBuilder sb = new StringBuilder();
        for (char c : word.toCharArray()) {
            int index = c - 'a';
            if (node.children[index] == null) {
                return word;
            }
            sb.append(c);
            node = node.children[index];
            if (node.isEnd) {
                return sb.toString();
            }
        }
        return word;
    }

    public String replaceWords(List<String> dictionary, String sentence) {
        root = new TrieNode();
        for (String word : dictionary) {
            insert(word);
        }
        String[] words = sentence.split(" ");
        StringBuilder sb = new StringBuilder();
        for (String word : words) {
            sb.append(search(word)).append(" ");
        }
        return sb.toString().trim();
    }

    @Override
    public Object solve(String[] inputJsonValues) {
        List<String> dictionary = jsonArrayToStringList(inputJsonValues[0]);
		String sentence = jsonStringToString(inputJsonValues[1]);
        return JSON.toJSON(replaceWords(dictionary, sentence));
    }
}
